File: monotonic-stack-queue/code/trapping-rain-leet-42-vertically-12-perfect.cpp
#include <vector>
#include <algorithm> // max()
using namespace std;

// 受 https://mp.weixin.qq.com/s/XyiYcDwEv3VW5Zs-WmRbDQ 启发，但是我自行想出，感觉很妙。可能是最佳解法。

    int trap(vector<int>& height) {
        int n = height.size();
        int maxh = 0; // 全局最高
        for (int i = 0; i < n; i++) {
            maxh = max(maxh, height[i]);
        }

        int sum = 0;
        int mh = 0; // 到目前为止，最高
        int i;
        for (i = 0; i < n && height[i] < maxh; i++) {
            mh = max(mh, height[i]);
            sum += (mh - height[i]) * 1;
        } // 以上遍历左部分

        int l = i; // l: 从左边数，第一个 maxh 的位置
        mh = 0;
        for (i = n - 1; i > l && height[i] < maxh; i--) {
            mh = max(mh, height[i]);
            sum += (mh - height[i]) * 1;
        } // 以上遍历右部分
        int r = i; // r: 从右边数，第一个 maxh 的位置
        for (i = l + 1; i <= r - 1; i++) {
            sum += (maxh - height[i]) * 1;
        } // 以上遍历中间部分
        return sum;
    }

File: monotonic-stack-queue/code/trapping-rain-leet-42-vertically-12-perfect.go
func trap(height []int) int {
    var maxh int = -1 // 全局最高柱子高度
    li, ri := -1, -1  // 最左、最右的 maxh 的下标
    for i, h := range height {
        if h > maxh {
            maxh = h
            li, ri = i, i // 同时更新最左边、最右边的 maxh 下标
        } else if h == maxh {
            ri = i // 只更新最右边的 maxh 下标
        }
    }

    sum := 0
    maxh_sofar := height[0]
    for i := 1; i < li; i++ {
        if h := height[i]; h < maxh_sofar { // 这个h的赋值，可以覆盖到下面的 else if
            sum += (maxh_sofar - h) * 1
        } else if h > maxh_sofar {
            maxh_sofar = h
        }
    }
    maxh_sofar = height[len(height) - 1]
    for i := len(height) - 1 - 1; i > ri; i-- {
        if h := height[i]; h < maxh_sofar {
            sum += (maxh_sofar - h) * 1
        } else if h > maxh_sofar {
            maxh_sofar = h
        }
    }
    for i := li + 1; i < ri; i++ {
        sum += (maxh - height[i]) * 1
    }
    return sum
}
